README.md


![R-CMD-check](https://github.com/microsud/microbiomeutilities/workflows/R-CMD-check/badge.svg)  


# About  
The `microbiomeutilities` R package is part of the [microbiome-verse](https://github.com/microbiome) tools that provides additional data handling and visualization support for the [microbiome R/BioC package](https://microbiome.github.io/tutorials/)   

Philosophy: "*Seemingly simple tasks for experienced R users can always be further simplified for novice users*"  

[Package website and online documentation](https://microsud.github.io/microbiomeutilities/)  
 
## Install microbiomeutilities    

```
install.packages("devtools")
devtools::install_github("microsud/microbiomeutilities")

```  
**Citation:**   

  o Leo Lahti, Sudarshan Shetty et al. (2017-2020). Tools for microbiome analysis in R. Version 2.1.28. URL: http://microbiome.github.com/microbiome      
  o Sudarshan A. Shetty, & Leo Lahti. (2020, October). `microbiomeutilities`: Utilities for Microbiome Analytics.   
  
  
The microbiome R package relies on the independently developed   
o [phyloseq](https://github.com/joey711/phyloseq) package and data structures for R-based microbiome analysis  developed by Paul McMurdie and Susan Holmes.  
o [ggplot2](http://ggplot2.org/) H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2009.  
o [tidyverse packages](https://www.tidyverse.org/).  

**Microbiome package website with step-wise tutorials**:  
URL: [http://microbiome.github.com/microbiome](http://microbiome.github.com/microbiome).  

**Tutorials**  
* [Microbiome R package tutorials](http://microbiome.github.io/microbiome/)  
* [Open and reproducible spring school 2018](https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/introduction.html)   
* [Tools Microbiome Anlaysis](https://microsud.github.io/Tools-Microbiome-Analysis/)   
 
### About the Author   
* [Web CV](https://microsud.github.io/gutmicrobe_webpage.io/)  
* [twitter](https://twitter.com/gutmicrobe)   
* [googlescholar](https://scholar.google.nl/citations?hl=en&user=Vahc6LUAAAAJ&view_op=list_works&sortby=pubdate)  
* [ORCID ID: 0000-0001-7280-9915](http://orcid.org/0000-0001-7280-9915)   


**MicrobiomeHD**   
The package provides access to a subset of studies included in the [MicrobiomeHD](https://zenodo.org/record/840333#.WuYuDoiFM2w) database from Duvallet et al 2017: [Meta-analysis of gut microbiome studies identifies disease-specific and shared responses. Nature communications](https://www.nature.com/articles/s41467-017-01973-8). These datasets are converted to phyloseq objects and can be directly used in R environment.  

### Datasets from:  
* Duvallet, Claire, et al. "Meta-analysis of gut microbiome studies identifies disease-specific and shared responses." Nature communications 8.1 (2017): 1784.   
* Son, J. et al. Comparison of fecal microbiota in children with autism spectrum disorders and neurotypical siblings in the simons simplex collection. PLoS ONE 10, e0137725 (2015).  
* Kang, D. W. et al. Reduced incidence of Prevotella and other fermenters in intestinal microflora of autistic children. PLoS ONE8, e68322 (2013).  
* Schubert, A. M. et al. Microbiome data distinguish patients with clostridium difficile infection and non-c. difficile-associated diarrhea from healthy controls. mBio 5, e01021–14–e01021–14 (2014).  
* Youngster, I. et al. Fecal microbiota transplant for relapsing clostridium difficile infection using a frozen inoculum from unrelated donors: a randomized, open-label, controlled pilot study. Clin. Infect. Dis. 58, 1515–1522 (2014).  
* Baxter, N. T., Ruffin, M. T., Rogers, M. A. & Schloss, P. D. Microbiota-based model improves the sensitivity of fecal immunochemical test for detecting colonic lesions. Genome Med. 8, 37 (2016).  
* Zackular, Joseph P., et al. "The gut microbiome modulates colon tumorigenesis." MBio 4.6 (2013): e00692-13.
* Zeller, G. et al. Potential of fecal microbiota for early-stage detection of colorectal cancer. Mol. Syst. Biol. 10, 766–766 (2014).
* Singh, P. et al. Intestinal microbial communities associated with acute enteric infections and disease recovery. Microbiome 3, 45 (2015).  
* Noguera-Julian, M. et al. Gut microbiota linked to sexual preference and hiv infection. EBioMedicine 5, 135–146 (2016).
Dinh, D. M. et al. Intestinal microbiota, microbial translocation, and systemic inflammation in chronic HIV infection. J. Infect. Dis. 211, 19–27 (2014).  
* Lozupone, C. A. et al. Alterations in the gut microbiota associated with hiv-1 infection. Cell Host Microbe 14, 329–339 (2013).
* Gevers, D. et al. The treatment-naive microbiome in new-onset crohn’s disease. Cell Host Microbe 15, 382–392 (2014).
* Zhang, Z. et al Large-scale survey of gut microbiota associated with MHE via 16s rRNA-based pyrosequencing. Am. J. Gastroenterol. 108, 1601–1611 (2013).  
* Wong, J. M. W., Souza, R. De, Kendall, C. W. C., Emam, A. & Jenkins, D. J. A. Colonic health: fermentation and short chain fatty acids. J. Clin. Gastroenterol. 40, 235–243 (2006).  
* Ross, M. C. et al. 16s gut community of the cameron county hispanic cohort. Microbiome 3, 7 (2015).
* Zupancic, M. L. et al. Analysis of the gut microbiota in the old order Amish and its relation to the metabolic syndrome. PLoS ONE 7, e43052 (2012).  
* Scher, J. U. et al. Expansion of intestinal Prevotella copri correlates with enhanced susceptibility to arthritis. eLife 2, e01202 (2013).  
* Alkanani, A. K. et al. Alterations in intestinal microbiota correlate with susceptibility to type 1 diabetes. Diabetes 64, 3510–3520 (2015).  
* Scheperjans, F. et al Gut microbiota are related to parkinson’s disease and clinical phenotype. Mov. Disord. 30, 350–358 (2014).

**NOTE:**  
The aim of this package is not to replace any of the other tools mentioned on this site.  
## Change log  

CHANGES IN VERSION 1.00.16 (2021-04-14)   
  o Fix `format_to_besthit` to return all slots present in input.   
  
CHANGES IN VERSION 1.00.15 (2021-01-21)   
  o Fix `plot_taxa_cv` to deal with taxa_are_rows==FALSE.   
  
CHANGES IN VERSION 1.00.14 (2021-01-21)   
  o Added utilities within `peak-methods`.   

CHANGES IN VERSION 1.00.12 (2021-01-18)  
  o Removed prevalence option `plot_taxa_heatmap`. 

CHANGES IN VERSION 1.00.12 (2021-01-18)    
  o Bug fix `format_to_besthit`.   
  
CHANGES IN VERSION 1.00.11 (2020-11-09)  
  o Small improvement in error handling.
  o Phyloseq slots to tibble
  o `add_refseq` for storing ASV sequences
  
CHANGES IN VERSION 1.00.10 (2020-11-02)  
  o Improve `get_group_abundances` documentation and code   

CHANGES IN VERSION 1.00.09 (2020-10-21)  
  o Added longitudinal page to website   
  o Added new function `plot_area`  
  o Added new function `plot_paired_abundances`  
  o Added new function `plot_spaghetti`   
  
CHANGES IN VERSION 1.00.08 (2020-10-17)  
  o Cosmetic updates   

CHANGES IN VERSION 1.00.07 (2020-10-15)  
  o Update to `plot_taxa_heatmap`    
  o Fixed `plot_abund_prev` options   
  o Added `plot_alpha_rcurve`   
CHANGES IN VERSION 1.00.06 (2020-10-13)  
  o Added new function dominant_taxa  
  o Removed plot_ternary due to clash between ggplot2 and ggtern    
  o Added new function find_samples_taxa    
  
CHANGES IN VERSION 1.00.05 (2020-10-11)  
  o Added new function `dominant_taxa`  
  o Added new function `get_group_abundances`    
  o removed `microbiome_pipeline`   
  
CHANGES IN VERSION 1.00.04 (2020-10-11)   
  o Added new function `plasticity`    
  o modified `theme_biome_utils`    
  
CHANGES IN VERSION 1.00.03 (2020-10-04)   
  o Added new function plot_ternary    
  o Deprecated plot_select_taxa    
  
CHANGES IN VERSION 1.00.02 (2020-10-04)   
  o Version tested with R version 4.0.2 (2020-10-04)         
  o Added new function `plot_listed_taxa`  
  o Deprecated `plot_select_taxa`  
  o Added option for half violin in boxplots     
  
CHANGES IN VERSION 1.00.01 (2020-10-03)
  o Added new function `plot_abund_prev`  
  o Added new function `simple_heatmap`  
  o Added new function `taxa_distribution`    
  o Added a custom theme `theme_biome_utils`    
  o Added `gghalves` to imports   
  o Fixed `microbiome_pipeline` report   
  
CHANGES IN VERSION 1.00.00 (2020-10-01)   
  o Version tested with R version 4.0.2 (2020-06-22)     
  o Fix typos in documentation  
  o Add prefix option to `format_to_besthit`  
  o Edited `phy_to_ldf` to speed up conversion  
  o Removed `format_phyloseq` function as it is redundant   
  o Speedup `taxa_summary` function  
  o Free up `pheatmap` option in `plot_taxa_heatmap`  
  o Updated for more info from `print_ps` output  
  o `plot_taxa_boxplot` now returns a faceted plot  
  o Added new function `plot_diversity_stats`  
  o R code styling `styler::tidyverse_style()` 
  
### References:
1. Callahan, B. J., McMurdie, P. J. & Holmes, S. P. (2017). Exact sequence variants should replace operational taxonomic units in marker gene data analysis. bioRxiv, 113597.  
2. Callahan, B. J., McMurdie, P. J., Rosen, M. J., Han, A. W., Johnson, A. J. A. & Holmes, S. P. (2016). DADA2: high-resolution sample inference from Illumina amplicon data. Nature methods 13, 581-583.  
3. Caporaso, J. G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F. D., Costello, E. K., Fierer, N., Peña, A. G., Goodrich, J. K. & Gordon, J. I. (2010). QIIME allows analysis of high-throughput community sequencing data. Nature methods 7, 335-336.  
4. Schloss, P. D., Westcott, S. L., Ryabin, T., Hall, J. R., Hartmann, M., Hollister, E. B., Lesniewski, R. A., Oakley, B. B., Parks, D. H. & Robinson, C. J. (2009). Introducing mothur: open-source, platform-independent, community-supported software for describing and comparing microbial communities. Applied and environmental microbiology 75, 7537-7541.  
Team, R. C. (2000). R language definition. Vienna, Austria: R foundation for statistical computing.  
 
### More useful resources:  
1. [Ben J. Callahan and Colleagues: Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses](https://f1000research.com/articles/5-1492/v2).   
2. [Comeau AM and Colleagues: Microbiome Helper: a Custom and Streamlined Workflow for Microbiome Research](http://msystems.asm.org/content/2/1/e00127-16)  
3.	MicrobiomeHD	[A standardized database of human gut microbiome studies in health and disease *Case-Control*](http://www.biorxiv.org/content/early/2017/05/08/134031)   
4.	Rhea	[A pipeline with modular R scripts](https://peerj.com/articles/2836/)  
5.	Phyloseq	[Import, share, and analyze microbiome census data using R](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217) 


inst/NEWS

Constant development is in process.  

CHANGES IN VERSION 1.00.16 (2021-04-14)   
  o Fix `format_to_besthit` to return all slots present in input.  
  
CHANGES IN VERSION 1.00.15 (2021-01-21)  
  o Fix `plot_taxa_cv` to deal with taxa_are_rows==FALSE.  
  
CHANGES IN VERSION 1.00.14 (2021-01-21)  
  o Added utilities within `peak-methods`.  
  
CHANGES IN VERSION 1.00.12 (2021-01-18)  
  o Removed prevalence option `plot_taxa_heatmap`.  
   
CHANGES IN VERSION 1.00.12 (2021-01-18)  
  o Bug fix `format_to_besthit`.  

CHANGES IN VERSION 1.00.11 (2020-11-09)  
  o Small improvement in error handling.
  o Phyloseq slots to tibble
  o `add_refseq` for storing ASV sequences
  
CHANGES IN VERSION 1.00.10 (2020-11-02)  
  o Improve `get_group_abundances` documentation and code   

CHANGES IN VERSION 1.00.09 (2020-10-21)  
  o Added longitudinal page to website   
  o Added new function plot_area  
  o Added new function plot_paired_abundances  
  o Added new function plot_spaghetti  
  
CHANGES IN VERSION 1.00.08 (2020-10-17)  
  o Cosmetic updates      
  
CHANGES IN VERSION 1.00.07 (2020-10-15)  
  o Update to `plot_taxa_heatmap`   
  o Fixed `plot_abund_prev` options  
  o Added `plot_alpha_rcurve`   
  
CHANGES IN VERSION 1.00.06 (2020-10-11)
  o Added new function dominant_taxa  
  o Removed plot_ternary due to clash between ggplot2 and ggtern  
  o Added new function find_samples_taxa    
  
CHANGES IN VERSION 1.00.05 (2020-10-11)
  o Added new function dominant_taxa  
  o Added new function get_group_abundances  
  o removed microbiome_pipeline  
  
CHANGES IN VERSION 1.00.04 (2020-10-11)
  o Added new function plasticity  
  o modified theme_biome_utils  
  
CHANGES IN VERSION 1.00.03 (2020-10-04)
  o Added new function plot_ternary  
  o Deprecated plot_select_taxa  
  
CHANGES IN VERSION 1.00.02 (2020-10-04)
  o Added new function plot_listed_taxa  
  o Deprecated plot_select_taxa  
  o Added option for half violin in boxplots     
  
CHANGES IN VERSION 1.00.01 (2020-10-03)
  o Added new function plot_abund_prev  
  o Added new function simple_heatmap  
  o Added new function taxa_distribution    
  o Added a custom theme theme_biome_utils    
  o Added gghalves to imports   
  o Fixed microbiome_pipeline report   
  
CHANGES IN VERSION 1.00.00 (2020-10-01)
  o Version tested with R version 4.0.2 (2020-06-22)     
  o Fix typos in documentation  
  o Add prefix option to format_to_besthit  
  o Edited phy_to_ldf to speed up conversion  
  o Removed format_phyloseq function as it is redundant   
  o Speedup taxa_summary function  
  o Free up pheatmap option in plot_taxa_heatmap  
  o Updated for more info from print_ps output  
  o plot_taxa_boxplot now returns a faceted plot  
  o Added new function plot_diversity_stats  
  o R code styling styler::tidyverse_style() 
  
CHANGES IN VERSION 0.99.02 (2020-02-02)
  o diversities to alpha in plot_alpha_diversities function
  o fix typos in documentation
  
CHANGES IN VERSION 0.99.01 (2018-04-09)
  o Cosmetic changes to report generator    
  
CHANGES IN VERSION 0.99.0 (2018-20-04)
  o Added print_ps (utility and alternative to print(ps))
  o Added taxa_summary, returns table with mean, max, median, standard deviation relative abundance
  o Use of abundance to extract otu_table to avoid bugs
  o Fixes for other bugs in examples and documentation  


R/print_ps.R
#' @title Overview of \code{\link{phyloseq-class}}
#' @description Prints an overview \code{\link{phyloseq-class}}.
#' @param x \code{\link{phyloseq-class}} object
#' @return Prints information about the \code{\link{phyloseq-class}} object.
#' @import utils
#' @export
#' @author Contact: Sudarshan A. Shetty \email{sudarshanshetty9@@gmail.com}
#' @examples
#' library(microbiomeutilities)
#' data("zackular2014")
#' pseq <- zackular2014
#' print_ps(pseq)
#' @keywords utilities

print_ps <- function(x) {
  comp <- NULL
  ave <- round(sum(sample_sums(x)) / nsamples(x), 2)
  #message(paste0("01] object is ", class(x)))
  message(paste0("01] ntaxa = ", ntaxa(x)))
  message(paste0("02] nsamples = ", nsamples(x)))
  message(paste0("03] nsamplesvariables = ", length(sample_variables(x))))
  message(paste0("04] nranks = ", length(colnames((tax_table(x))))))
  message(paste0("05] Min. number of reads = ", min(sample_sums(x))))
  message(paste0("06] Max. number of reads = ", max(sample_sums(x))))
  message(paste0("07] Total number of reads = ", sum(sample_sums(x))))
  message(paste0("08] Average number of reads = ", ave))
  message(paste0("09] Median number of reads = ", median(sample_sums(x))))
  message(paste0("10] Sparsity = ", length(which(abundances(x) ==
    0)) / length(abundances(x))))
  message(paste0(
    "11] Number of singletons = ",
    length(taxa_sums(x)[taxa_sums(x) <=
      1])
  ))
  message(paste0(
    "12] % of taxa that are singletons 
  (i.e. exactly one read detected across all samples) = ",
    mean(taxa_sums(x) == 1) * 100
  ))
}
